Save space saving sizes of dynamic fields only

twiggler opened the pull request:
Currently,  actual sizes of all fields are stored inside the `_sizes` private member. This can cause high memory usage in some applications. 

This PR only stores the sizes of dynamic fields. Since most fields are static, we significantly reduce memory usage, utilities like `dumpstruct` still work as intended. 

Alternative solution considered: Remove `_sizes` completely. However, no alternative way has been found to reconstruct the number of bytes read. This would result in loss of the pretty colors in dump_struct, for example

twiggler committed:
feature: Save space saving sizes of dynamic fields only

twiggler committed:
Add sizes property

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -403,6 +402,18 @@ def __repr__(self) -> str:
         return f"<{self.__class__.__name__} {' '.join(values)}>"
 
+    @cached_property
+    def sizes(self) -> None:
```
We can't really add a `sizes` property since it will clash with structures that have a field names `sizes`. Which is why almost everything related to these classes is either `_` or `__` 😄.

twiggler replied:
Ah yes, a similar mistake as the  `default` method IIRC. 

I renamed to `__field_sizes__`

JSCU-CNI replied:
Wouldn't it make more sense to keep using `_sizes` instead of `__field_sizes__`?

Schamper replied:
@JSCU-CNI we've decided to rename them to `__values__` and `__sizes__`. This is more in line with the rest of the "internal attribute" naming and formalizes it a bit more.

twiggler committed:
Rename _ values  to __values__ and sizes to __sizes__

twiggler committed:
enhancement: make values return a proxy dict

twiggler committed:
optimization: cache static sizes on class

twiggler commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+    def __init__(self, struct: Structure):
+        self._struct: Structure = struct
+        # Unfortunately, we need to support code that adds attributes dynamically :/
+        self._fields: set[str] = set(iter(self._struct.__class__.fields))
```
Alternatively, we could use a ChainMap here over field names and extra fields. This would be more memory efficient, but slightly slower when iterating (since chainmap needs to preserve key order).

Schamper replied:
What do you mean extra fields? You can't dynamically add fields to an already parsed structure.

twiggler replied:
See above, although "extra fields" should have been "extra attributes"

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+
+    def __init__(self, struct: Structure):
+        self._struct: Structure = struct
+        # Unfortunately, we need to support code that adds attributes dynamically :/
```
We do? To a structure object? Where does such code live?

twiggler replied:
For example [here](https://github.com/fox-it/dissect.etl/blob/3219d88bb6c6668fedb40a5394519c3cef54ee89/dissect/etl/headers/event.py#L55)

Since  ETL has been programmed against an older version of cstruct, I asked MiauwKeru  and  he confirmed that adding attributes should be possible, even if the structure has no corresponding field.

Moreover, we don´t check when setting an attribute to see if it has a corresponding field, so I gathered that this is expected behavior.

Personally, I think adding new attributes this way is is a bad idea. 

Should I remove this capability? 
And perhaps add a check to `setattr` of structure that we don´t create attributes without corresponding fields?

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+        return len(self._fields)
+
+    def __repr__(self) -> str:
+        return repr(self._struct)
```
Shouldn't the `repr` be more like a dictionary?

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+        self._fields: set[str] = set(iter(self._struct.__class__.fields))
+
+    def __getitem__(self, key: str) -> Any:
+        return getattr(self._struct, key)
```
```suggestion
        if key in self._fields:
            return getattr(self._struct, key)
        raise KeyError(key)
```
?

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        setattr(self._struct, key, value)
+        self._fields.add(key)
```
```suggestion
        if key in self._fields:
            setattr(self._struct, key, value)
        raise KeyError(key)
```
?

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +407,46 @@ def __repr__(self) -> str:
+
+    def __delitem__(self, key: str) -> None:
+        del self._struct[key]
+        self._fields.discard(key)
```
Don't think this should be allowed?

twiggler replied:
Yeah it seems like a bad idea, but we don´t guard against deleting fields from structures so I thought it should be possible.

Shall I disable?

twiggler committed:
fix: Do not set attributes with no corresponding field.

The implicit interface does not allow this.
Added a docstring to make it more explicit.

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +411,49 @@ def __repr__(self) -> str:
+
+    @property
+    def __values__(self) -> MutableMapping[str, Any]:
+        return StructureValuesProxy(self)
```
```suggestion
    @property
    def __values__(self) -> MutableMapping[str, Any]:
        return StructureValuesProxy(self)

    @property
    def __sizes__(self) -> Mapping[str, int | None]:
        return ChainMap(self.__class__.__static_sizes__, self.__dynamic_sizes__)
```
Just for consistency sake.

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +411,49 @@ def __repr__(self) -> str:
+
+    def __repr__(self) -> str:
+        items = (f"{key!r}: {value!r}" for (key, value) in self.items())
+        return "{" + ", ".join(items) + "}"
```
```suggestion
        return repr(dict(self))
```
Would this work?

twiggler replied:
Tested fine, it is more elegant

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +411,49 @@ def __repr__(self) -> str:
+        if key in self:
+            setattr(self._struct, key, value)
+        else:
+            raise KeyError(key)
```
```suggestion
        raise KeyError(key)
```

Schamper commented on dissect/cstruct/types/structure.py:
```diff
@@ -401,6 +411,49 @@ def __repr__(self) -> str:
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if key in self:
+            setattr(self._struct, key, value)
```
```suggestion
            return setattr(self._struct, key, value)
```

twiggler replied:
meh :/

Schamper commented on tests/test_compiler.py:
```diff
@@ -337,6 +318,7 @@ def test_generate_fields_dynamic_after_bitfield(cs: cstruct, TestEnum: Enum, oth
         Field("b", _type, 4),
         Field("c", cs.char["size"], offset=3),
     ]
+    fields[3].type.dynamic = True   # cs.char["size"] is a hack so patch dynamic
```
What do you mean with it being a hack?

twiggler replied:
AFAIK   `Field("c", cs.char["size"], offset=3)` does not set dynamic is True. 

Maybe mock is a better term

Schamper replied:
Looks like faulty usage of this. `_make_array` expects an integer size or an `Expression` here (from the syntax parser). On the one hand I believe this should be allowed usage [since it's possible to add fields yourself from code](https://github.com/fox-it/dissect.cstruct/blob/e990b5301ffa6c739cdbd954baeb70d3a970a362/dissect/cstruct/types/structure.py#L354-L359). I think it can be solved by simply wrapping the string in an `Expression` in `_make_array`.

Or we don't allow this, in which case we should turn this "mock" more proper by making it `cs.char[Expression("size")]`.

twiggler replied:
When we creating fields programmatically I prefer the more explicit `cs.char[Expression("size")]`. 

However, the Expression constructor also required a cstruct instance. Because the cstruct instance is only required during evaluation, I made it a parameter of `Expression::evaluate`

twiggler committed:
Use expression when programatically creating fields.

twiggler committed:
Process minor code style remarks

Schamper commented on tests/test_compiler.py:
```diff
@@ -377,6 +358,7 @@ def test_generate_fields_dynamic_before_bitfield(cs: cstruct, TestEnum: Enum, ot
         Field("b", TestEnum, 4),
         Field("c", cs.char["size"], offset=3),
     ]
+    fields[3].type.dynamic = True   # cs.char["size"] is a hack so patch dynamic
```
Fix here too?

twiggler committed:
fix: dynamic size in other compiler test

JSCU-CNI commented:
Could this change be documented, e.g. using a `CHANGELOG` file entry?

twiggler committed:
Add changelog

Schamper commented on CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+## Library
+- Optimize storage of field sizes.
+- Rename `_sizes` property of `Structure` to `__sizes__`. 
+- Rename `_values` property of `Structure` to `__values__`.
```
```suggestion
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).

## [Unreleased]

### Changed

- Optimize storage of field sizes
- Rename `_sizes` property of `Structure` to `__sizes__`.
- Rename `_values` property of `Structure` to `__values__`

[unreleased]: https://github.com/fox-it/dissect.cstruct/compare/4.5...HEAD
```
I looked online for some examples of nice changelog formats and stumbled upon this, maybe we can try it? What do you think.

I'm thinking the comparison URLs are a bit superfluous with what's already available in https://github.com/fox-it/dissect.cstruct/releases, so maybe we should leave that out. Also maybe consider adding the previous changelogs in here too?

twiggler replied:
> What do you think.

I don´t have a strong opinion about the format. The format you propose looks fine. With regard to the contents, it think it is important that entries are written with the user in mind.

I do wonder how the changelog compares and contrasts with the github releases functionality in *general* (i.e. https://github.com/fox-it/dissect.cstruct/releases). It seems they share the same purpose? Maybe we can automate keeping them in sync.

> I'm thinking the comparison URLs are a bit superfluous with what's already available in https://github.com/fox-it/dissect.cstruct/releases, so maybe we should leave that out. 

I think they look ugly, in favor of dropping them.

> Also maybe consider adding the previous changelogs in here too?

I can add the changes starting from 4.0. Version four seems to be a complete rewrite, and I don´t have enough context to write a meaningful changelog beyond. Personally I wouldn´t bother though.

twiggler committed:
tweak changelog
